engine/src/main/resources/assets/shaders/chunk_frag.glsl

```diff
@@ -92,34 +92,41 @@ void main() {
     vec3 normalOpaque = normal;
     float shininess = 0.0;
 
-#if defined (NORMAL_MAPPING)
+#if defined (NORMAL_MAPPING) || defined (PARALLAX_MAPPING)
     // TODO: Calculates the tangent frame on the fly - this is absurdly costly... But storing
     // the tangent for each vertex in the chunk VBO might be not the best idea either.
-    vec3 dp1 = dFdx(vertexProjPos.xyz);
-    vec3 dp2 = dFdy(vertexProjPos.xyz);
-    vec2 duv1 = dFdx(gl_TexCoord[0].xy);
-    vec2 duv2 = dFdy(gl_TexCoord[0].xy);
-
-    vec3 dp2perp = cross(dp2, normal);
-    vec3 dp1perp = cross(normal, dp1);
-    vec3 tangent = dp2perp * duv1.x + dp1perp * duv2.x;
-    vec3 binormal = dp2perp * duv1.y + dp1perp * duv2.y;
-
-    float invMax = inversesqrt(max(dot(tangent,tangent), dot(binormal,binormal)));
-    mat3 tbn = mat3(tangent * invMax, binormal * invMax, normal);
+    // Screen-space coordinates happen to be a coordinate system to which it's possible to relate both view-space and texture coordinates, but are themselves irrelevant.
+    mat2x3 screenToView = mat2x3(dFdx(vertexViewPos.xyz), dFdy(vertexViewPos.xyz));
+    mat2   screenToUv   = mat2  (dFdx(gl_TexCoord[0].xy), dFdy(gl_TexCoord[0].xy)) / TEXTURE_OFFSET;
+    //mat2 uvToScreem = inverse(screenToUv);
+    //inverse is not available in GLSL 1.20, so calculate it manually:
+        float det = screenToUv[0][0] * screenToUv[1][1] - screenToUv[1][0] * screenToUv[0][1];
+        mat2 uvToScreen = mat2(vec2(screenToUv[1][1],-screenToUv[0][1]),vec2(-screenToUv[1][0],screenToUv[0][0])) / det;
+    mat2x3 uvToView = screenToView * uvToScreen;
+    mat3 tbn = mat3(uvToView[0], uvToView[1], normal);
```

Comment:
Hmmmm. I know what the three vectors in this matrix are and I know "tbn" is a frequently used variable name in the field of computer graphics and shaders. Nevertheless, I wonder if we could find a more meaningful name for it. 

I guess this matrix alone is not meant as a tool to convert from one space to another, it just represent a space, right? If that's the case I guess I'd be happy simply with "tbnSpace". Or does it have other purposes?

Reply:
As it's an extension of uvToView, it maps vectors from the UV and normal basis to view-space coordinates, but that first basis doesn't have a suitable descriptive name I can think of. The name tangent-binormal-normal is definitely inappropriate here though.